### PR TITLE
Remove tensorflow from h1st dependencies

### LINKED ISSUE #160: Remove tensorflow from h1st dependencies
## What ?
Currently, `h1st` require `tensorflow = ">=2.10.0"` as one of it dependencies but we should remove `tensorflow` from h1st dependencies list.

### `tensorflow` is not easy to install correctly on all architectures/platforms

Tensorflow is a machine learning framework that support multiple OS (Windows, Mac, Linux) and most of the Python version. But due to performance reason, it also have to support many different hardware acceleration method (CPU/GPU/TPU/NPU) and different compute architecture (x86, ARM, IBM Power). To take advantage of all these different platform, it have to depend on hardware specific framework and SDK such as: CUDA, CuDNN, TensorRT (for NVIDIA GPU) or MKL-DNN, OpenAPI (for Intel CPU), ROCm (for AMD CPU/GPU/APU), Metal, CoreML (for Apple CPU/GPU/Neural Engine), ... These SDK often get dynamically linked to the Tensorflow binary and require another level of dependencies management on it own so it can be utilize properly by Tensorflow.

Build, maintain, test and release all of these variation of the package is already a huge challenge for the Tensorflow team, and they already had to defer the work to multiple third party such as AWS, Intel and NVIDIA. This make it very difficult to trivially install Tensorflow on multiple platform, just take a quick look on this [official installation instruction](https://www.tensorflow.org/install/pip#linux) from Tensorflow. You can already see that each platform have it own installation command, some use `conda`, some use `pip`, some even have it's own separate python package.

Since it is crucial for the user to install the correct variation of Tensorflow so they can take full advantage of their hardware. And there is no way for `poetry` (our package manger) to accommodate all of that variation, we should not force `h1st`'s user to go with a generic `tensorflow>=2.10.0` requirement.

### `tensorflow` is a heavy package that are huge in size

Looking at the [PyPI pages](https://pypi.org/project/tensorflow/#files) that track all of `tensorflow` wheel files, we can see that in the majority of the wheel files for all platforms and python version have size bigger than 400 MB, if we count all the other dependencies that `tensorflow` require, it can easily go up to 500 MB. After unpack for installation, it could took close to a whole GB of space in most cases.

### `tensorflow` does not get use very often, even internally by Aitomatic

Most of our work use `scikit-learn` and not `tensorflow`. For more complex model, `pytorch` are increasingly being adopted by the world and our own engineer instead of `tensorflow`. So why make `tensorflow` a mandatory dependencies on every single project that use `h1st` when it just sitting there as dead weight, costing us more in time, disk space and network transfer cost.

### What if the user want to use their own `tensorflow` version ?

Many legacy ML system still uses `tensorflow<=1.14` so by requiring `tensorflow = ">=2.10.0"` in `h1st`, we will lose compatibility with all of those system and potential user. This will slow down the adoption of H1st.

### `tensorflow` can be safely remove from `h1st` codebase altogether

If we look for [where `tensorflow` are being used in `h1st` codebase](https://github.com/h1st-ai/h1st/search?q=tensorflow&type=code), you can quickly see that it is only being use for saving and loading MLModel that originate from Tensorflow.

This saving and loading logic [can be implemented](https://github.com/h1st-ai/h1st/blob/6471b8657e49a09a13b3d6d06f9aaa446a559571/h1st/model/model.py#L66) by `h1st` user to add support for any type of model that they might need to save and load along side with each h1st's MLModel. We already did this [internally](https://github.com/aitomatic/fixed-net/blob/1975be9e045ab324a86b41c4d8c48f003986cee3/ai/models/fish_detection.py#L90) to support pytorch model as well. It is pretty easy to do ;)

## How ?
We should remove `tensorflow` from h1st dependencies list, along side with any saving and loading code for `tensorflow`

## Any alternatives ?
In theory one can provide a `conda` repo with every variation of the package similar to what pytorch did [here](https://pytorch.org/get-started/locally/#start-locally)

Even then:
- It will still be limited by what platform and architecture that `conda` can support
- It take massive amount of time and resource to do. Both Google and Meta with their massive amount of resource already gave up on build, maintain and release every variation by themself.

I think we should consider this alternatives impractical and shouldn't bother trying this way. Each user should decide how they want to install `tensorflow` for themself and how they want to use it with `h1st`.

## **Additional context**
Same suggestion was made in #105

### PR DESCRIPTION
**Pull request by vuonghoainam**
This PR fixes #160

### CODE REVIEW

**Comment by nubs01** [comment]
I highly advise against dropping tensorflow support. Customers will still need to be able to persist tensorflow models. There is a way to implement this support so that if tensorflow is installed it will work, without requiring that tensorflow be installed

**Review by ctn — APPROVED** [review]
lgtm
